### PR TITLE
609 enable lusd collateralization

### LINKED ISSUE #609: Enable LUSD Collateralization
In the same fashion as [FRAX](https://docs.frax.finance/minting-and-redeeming#minting)
we want a minting and redeeming uAD mechanism where users can deposit LUSD or sUSD and get uAD in return.
We don't require any gov token to mint. 
uAD price should be greater than $1.005 to allow for minting otherwise users must purchase from the market

### PR DESCRIPTION
**Pull request by hashedMae**
Resolves #609

### CODE REVIEW

**Comment by 0x4007** [comment]
Is there an issue with the CI? I'm on mobile so my capabilities are limited. 

Could you request reviews from all the Solidity developers? It isn't easy for me to on the iOS client.  

@zgorizzo69 @0xcodercrane @rndquu @Steveantor

**Comment by hashedMae** [comment]
> Is there an issue with the CI? I'm on mobile so my capabilities are limited.


"Compare Test Coverage" is failing, but I have no idea what that is for.

**Comment by 0x4007** [comment]
> > Is there an issue with the CI? I'm on mobile so my capabilities are limited.
> 
> 
> 
> 
> 
> "Compare Test Coverage" is failing, but I have no idea what that is for. 

Can you verify that the percentage of test coverage has not decreased with this pull request?

That is the intended purpose of that CI.

**Comment by hashedMae** [comment]
It doesn't even get that far. Just `Exit code 1`

**Comment by rndquu** [comment]
> It doesn't even get that far. Just `Exit code 1`

I'll check it in few hours

**Comment by rndquu** [comment]
@hashedMae 

The CI action "Compare Test Coverage" fails due to decreased test coverage.  

If you check the "Print coverages" step [here](https://github.com/ubiquity/ubiquity-dollar/actions/runs/4834823677/jobs/8684367288?pr=627) you may see the development branch test coverage and PR's test coverage:
```
Development branch coverage: 38.17
PR branch coverage: 35.82
```

This [PR](https://github.com/ubiquity/ubiquity-dollar/pull/631) makes the coverage error more verbose

Pls add more tests for the "LUSD collateralization" feature

**Review by 0x4007 — CHANGES_REQUESTED** [review]
Token symbols should never be referenced in any identifiers. See more information [here](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/CONTRIBUTING.md).

**Review by 0x4007 — CHANGES_REQUESTED** [review]
Token symbols should never be referenced in any identifiers. See more information [here](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/CONTRIBUTING.md).

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol:1]: Token symbols should never be referenced in any identifiers. See more information [here](https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/CONTRIBUTING.md).

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function setCollateralWrappedEtherOracle(
```

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function buyBackGovernance(
```

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function recollateralizeDollar(
```

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function redeemDollar(
```

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function redeemFractionalDollar(
```

I'm not sure what this is

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: What does this mean?

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]:
```suggestion
    function mintFractionalDollar(
```

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol:1]: It looks like the variables are snake cased. They should be camel cased?

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]:
```suggestion

```

File: packages/contracts/test/diamond/DiamondTest.t.sol
Comment (0x4007) [inline packages/contracts/test/diamond/DiamondTest.t.sol:15]: Any idea how to handle this dynamically

File: packages/contracts/test/diamond/DiamondTestSetup.sol
Comment (0x4007) [inline packages/contracts/test/diamond/DiamondTestSetup.sol:509]: Perhaps we should be descriptive like `Ubiquity3PoolFacet` etc

File: packages/contracts/test/diamond/DiamondTestSetup.sol
Comment (0x4007) [inline packages/contracts/test/diamond/DiamondTestSetup.sol:512]:
```suggestion
            "CreditNftManagerFacet",
```

**Review by 0x4007 — CHANGES_REQUESTED** [review]

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (0x4007) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: Can you elaborate on why these are separate mint functions? Fractional and algorithmic mint.

Comment (zgorizzo69) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: outdated

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (zgorizzo69) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: outdated

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (zgorizzo69) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: outdated

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (zgorizzo69) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: outdated

File: packages/contracts/test/diamond/DiamondTest.t.sol
Comment (zgorizzo69) [reply packages/contracts/test/diamond/DiamondTest.t.sol:15]: not really necessary IMO it is for testing only

File: packages/contracts/test/diamond/DiamondTestSetup.sol
Comment (zgorizzo69) [reply packages/contracts/test/diamond/DiamondTestSetup.sol:512]: it is only used by test and not part of that pr

File: packages/contracts/test/diamond/DiamondTestSetup.sol
Comment (zgorizzo69) [reply packages/contracts/test/diamond/DiamondTestSetup.sol:509]: only part of the test

File: packages/contracts/src/dollar/interfaces/IUbiquityPool.sol
Comment (zgorizzo69) [reply packages/contracts/src/dollar/interfaces/IUbiquityPool.sol]: outdated

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:24]: The `ubiquity.contracts.ubiquity` prefix seems redundant.

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]:
```suggestion
            "Ubiquity Dollar Token value must be 1 USD or greater to mint"
```

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]:
```suggestion
            "Ubiquity Dollar Token value must be less than 1 USD to redeem"
```

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]: This is a vague error perhaps it could be rewritten?

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]: What is `Curve3`? If it is referring to the Curve3Pool I think we should write it out like that!

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:305]:
```suggestion
    function calculateCollateralRedeemAmount(
```

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]:
```suggestion
    function getDollarPriceUsd()
```

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [inline packages/contracts/src/dollar/libraries/LibUbiquityPool.sol]:
```suggestion
    function getCollateralPriceCurve3Pool(
```

**Review by 0x4007 — CHANGES_REQUESTED** [review]
I'm good to merge after these last small adjustments. Thanks for implementing the code changes from before.

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (0x4007) [reply packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:305]: I don't think we should shorten any variable names in order to increase code clarity.

File: packages/contracts/src/dollar/libraries/LibUbiquityPool.sol
Comment (Steveantor) [reply packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:24]: we need a unique name for each contract's storage position to avoid collisions

Comment (0x4007) [reply packages/contracts/src/dollar/libraries/LibUbiquityPool.sol:24]: I understand this. I still stand behind my statement.

**Review by 0x4007 — APPROVED** [review]

**Comment by 0x4007** [comment]
There are a few unresolved conversations. Do you know how to find them conveniently? 
image

**Comment by zgorizzo69** [comment]
> There are a few unresolved conversations. Do you know how to find them conveniently? image

they are outdated so not relevant anymore

**Comment by 0x4007** [comment]
You need to resolve all outstanding comments before GitHub merges.

**Comment by 0x4007** [comment]
image

Not sure about this [curve3PoolTokenAddress](https://github.com/ubiquity/ubiquity-dollar/blob/2bffaf01d209c1a167c2e689711699daf92012ba/packages/dapp/components/lib/hooks/contracts/useManagerManaged.tsx#L68)

**Comment by zgorizzo69** [comment]
> image
> 
> Not sure about this [curve3PoolTokenAddress](https://github.com/ubiquity/ubiquity-dollar/blob/2bffaf01d209c1a167c2e689711699daf92012ba/packages/dapp/components/lib/hooks/contracts/useManagerManaged.tsx#L68)

https://github.com/ubiquity/ubiquity-dollar/blob/8db0891b6488e8262ec7c269889e6a7577679b44/packages/contracts/src/dollar/facets/ManagerFacet.sol#L252 :monocle_face: strange have you tried clean up and regenerate types again ?

**Pull request merged**
